### PR TITLE
Correcting cholesky_backward

### LINKED ISSUE #137284: Incorrect Cholesky backward grad due to incorrect symmetrization
### 🐛 Describe the bug

I believe `cholesky_backward` implementation is incorrect due to the false but well-intentioned assumption that $\overline{\Sigma}=\frac{S + S^\top}{2}$ is the backward gradient for $\Sigma$ once we get the backward gradient $S$, but the math tells us that if $dy = Tr(Sd\Sigma)$ then  $\overline{\Sigma}=S + S^\top -diag(S)$.

In the current implementation, the former incorrect approach is used. 
We can prove that $S = L^{-\top}\Phi(L^\top \overline{L})L^{-1}$ (See [paper](https://arxiv.org/abs/1602.07527)) where $\Phi(B)=tril(B) - \frac{1}{2}diag(B)$ (take the lower triangular and halves the diagonal).

## Current implementation
To make sense of the current implementation, it uses
$\overline{\Sigma}= L^{-\top}\left[\frac{\Phi(L^\top \overline{L}) + \Phi(L^\top \overline{L})^\top}{2}\right]L^{-1}$
Using $M = L^\top \overline{L}$:
$$\frac{\Phi(L^\top \overline{L}) + \Phi(L^\top \overline{L})^\top}{2}=\frac{1}{2}\left[tril(M) + tril(M)^\top -diag(M)\right]$$
with $A=tril(M)$ it's
$$\frac{1}{2}\left[A + A^\top -diag(A)\right] =\left[A + tril(A,-1)^\top\right]$$ (always true if A is lower triangular)
## The math way
Compute $S$ from above and return $S + S^\top -diag(S)$. I couldn't come up with a more concise way using the tril(A,-1) because of diag(S).

## Examples
Let's see some numbers.

Consider the symmetric PD matrix with $s,u>0$ and $u>\frac{t^2}{s}$
$$\Sigma = [[s , t],  [t , u]]$$

Cholesky decomposition is:
$$L = [[\sqrt{s} , 0 ],[\frac{t}{\sqrt{s}} , \sqrt{u - \frac{t^2}{s}}]]$$

Let's take s,t,u = 1,1,2 and backwards $L_{1,2}$ , we expect $\overline{t}=\frac{1}{\sqrt(s)}=1$.
What we get:
```python
import torch
## Current result
S = torch.tensor([[1.,1.],[1.,2.]], requires_grad=True)
A = torch.tril(torch.ones((2,2)),-1)
# >> A
# tensor([[0., 0.],
#         [1., 0.]])
L = torch.linalg.cholesky(S)
loss = (L*A).sum()
loss.backward(reta)
# >> S.grad[0,1]
# tensor(0.5000)

## What we should get if we manually compute L =cholesky(S)
S = torch.tensor([[1.,1.],[1.,2.]], requires_grad=True)
manual_L = torch.stack([torch.cat([S[0][:1]**.5, torch.tensor([0.0])]),
    torch.cat([S[0][1:]/S[0][:1]**.5, (S[1][1:] - S[0][1:]**2/S[0][:1])**.5])])
true_loss = torch.sum(e_L*A)
true_loss.backward()
# >> S.grad[0,1]
# tensor(1.)
```

```python
import torch
torch.manual_seed(2024)
C=torch.randn(4,4)/4.0
S = C.t().mm(C)
S.requires_grad_()
A = torch.randn(4,4)/4.0
loss = (A*torch.linalg.cholesky(S)).sum()
loss.backward()
S.grad
## current grad
tensor([[4.8560, 5.0956, 4.2147, 2.6087],
        [5.0956, 4.6948, 3.8186, 2.2791],
        [4.2147, 3.8186, 3.4634, 1.8416],
        [2.6087, 2.2791, 1.8416, 0.7801]])

## proposed pull request
tensor([[ 4.8560, 10.1912,  8.4293,  5.2173],
        [10.1912,  4.6948,  7.6372,  4.5582],
        [ 8.4293,  7.6372,  3.4634,  3.6832],
        [ 5.2173,  4.5582,  3.6832,  0.7801]])
```

## TLDR
Incorrect `cholesky_backward` due to incorrect symmetrization of the backward sensitivity using `(S + S^H)/2` which halves the non-diagonal elements. The math tells us that we should mirror the non-diagonal sensitivities while keeping the diagonal ones using `S + S^H - diag(S)`.

### Versions

Collecting environment information...
PyTorch version: 2.4.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Gentoo Linux (x86_64)
GCC version: (Gentoo 13.3.1_p20240614 p17) 13.3.1 20240614
Clang version: 18.1.8
CMake version: version 3.28.5
Libc version: glibc-2.39

Python version: 3.12.3 (main, Jun 11 2024, 01:10:37) [GCC 13.2.1 20240210] (64-bit runtime)
Python platform: Linux-6.6.38-gentoo-x86_64-x86_64-11th_Gen_Intel-R-_Core-TM-_i7-11700K_@_3.60GHz-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 550.107.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               16
On-line CPU(s) list:                  0-15
Vendor ID:                            GenuineIntel
Model name:                           11th Gen Intel(R) Core(TM) i7-11700K @ 3.60GHz
CPU family:                           6
Model:                                167
Thread(s) per core:                   2
Core(s) per socket:                   8
Socket(s):                            1
Stepping:                             1
CPU(s) scaling MHz:                   81%
CPU max MHz:                          5000.0000
CPU min MHz:                          800.0000
BogoMIPS:                             7202.00
Flags:                                fpu ....
Virtualization:                       VT-x
L1d cache:                            384 KiB (8 instances)
L1i cache:                            256 KiB (8 instances)
L2 cache:                             4 MiB (8 instances)
L3 cache:                             16 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Vulnerable: No microcode
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.1.1
[pip3] pytorch-lightning==2.4.0
[pip3] torch==2.4.1
[pip3] torch_cluster==1.6.3+pt24cu121
[pip3] torchmetrics==0.11.4
[pip3] torchvision==0.19.1
[pip3] triton==3.0.0
[conda] No relevant packages

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Varal7 @xmfan @jianyuh @mruberry @walterddr @xwang233 @Lezcano

### PR DESCRIPTION
**Pull request by ayghri**
incorrect cholesky_backward due to halving the non-diagonal elements when making the backward sensitivities symmetric.

Fixes #137284

### CODE REVIEW

**Review by lezcano — CHANGES_REQUESTED** [review]
cf. https://github.com/pytorch/pytorch/issues/137284#issuecomment-2393243777

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -1924,7 +1924,7 @@ Tensor cholesky_jvp(const Tensor& dA, const Tensor& L, bool upper) {
   //               = sym(L^{-1}dL)
   // where sym(X) = X + X^H
   // A short computation gives that the inverse of sym is given by
-  // \pi(X) = X.tril() - 0.5*diag(X)
+  // \pi(X) = X.tril() - 0.5*diag(X) when X is lower triangular,
```
Comment (lezcano) [inline torch/csrc/autograd/FunctionsManual.cpp:1927]: This is not correct. `sym : Tril(n) -> Sym(n)` so `\pi : Sym(n) -> Tril(n)` and `X` here is symmetric.

Comment (ayghri) [reply torch/csrc/autograd/FunctionsManual.cpp]: I've reverted the changes to this comment.

**Comment by lezcano** [comment]
The tests are not passing
